ams/server/src/main/java/com/netease/arctic/server/table/TagConfiguration.java

```diff
@@ -73,6 +95,32 @@ public String propertyName() {
      * offset is set to be 5 min, the idea trigger time is 2022-08-08 00:05:00.
      */
     public abstract long getTagTriggerTime(LocalDateTime checkTime, int triggerOffsetMinutes);
+
+    public abstract LocalDateTime normalizeToTagTime(long triggerTime, int triggerOffsetMinutes);
+
+    /**
+     * calculate the end of the period adjusted with a specified offset.
+     *
+     * <p>The calculation takes into account the provided period size and offset. It first
+     * calculates an offset timestamp, which is essentially `checkTime` minus offset minutes. Then
+     * it calculates the delta of the division of that timestamp by the period size, then applies
+     * the delta to the original timestamp to get the period start timestamp.
+     */
+    protected Long getPeriodEndWithOffset(
```

Comment:
This implementation is a little complex. I attempted to simplify it using the existing API of LocalDateTime, as shown below.

https://github.com/huyuanfeng2018/amoro/pull/1/files

Reply:
> This implementation is a little complex. I attempted to simplify it using the existing API of LocalDateTime, as shown below.
> 
> https://github.com/huyuanfeng2018/amoro/pull/1/files
 
Well, this does make that logic simpler, but it is because our current `Period` only exists DAILY and HOURLY. Just imagine, what if the `Period` here is 2 hours?

Reply:
I suggest keeping the current code simple before introducing configurations like `two-hourly`.

Furthermore, I don't think introducing `Period` like `two-hourly`, `four-hourly`, `half-day` is a generic solution. We can discuss  how to support such requirements in the future useless you currently require it.

Reply:
> I suggest keeping the current code simple before introducing configurations like `two-hourly`.
> 
> Furthermore, I don't think introducing `Period` like `two-hourly`, `four-hourly`, `half-day` is a generic solution. We can discuss how to support such requirements in the future useless you currently require it.

ok. I adjusted some code to make it more readable